### PR TITLE
Changes in feature/error-handling-staging 

### PR DESCRIPTION
**Pull request by balos1**
These are just the changes in feature/error-handling-staging that the other branches are based on.

### CODE REVIEW

File: src/sundials/sundials_errors.c [hunk 1]
```diff
@@ -77,10 +75,38 @@ void SUNAbortErrHandlerFn(int line, const char* func, const char* file,
...
+{
+  va_list ap;
+  char* log_msg       = NULL;
+  char* file_and_line = NULL;
+
+  va_start(ap, err_code);
+
+  file_and_line = sunCombineFileAndLine(__LINE__, __FILE__);
+  sunCreateLogMessage(SUN_LOGLEVEL_ERROR, 0, file_and_line,
+                      __func__, "The SUNDIALS SUNContext was corrupt or NULL when an error occurred. As such, error messages have been printed to stderr.",
```
Comment (drreynolds) [inline src/sundials/sundials_errors.c:97]: Is it really *always* the case that if `SUNGlobalFallbackErrHandler` is called, then the SUNDIALS SUNContext was corrupt or NULL when the error occurred, or is it possible to use this fallback function just because no other error handlers were specified?  If this could be used in any other situation, then I recommend that this line be adjusted.

Comment (balos1) [reply src/sundials/sundials_errors.c:97]: It will only ever be called if the `SUNContext` is `NULL` or the object holding the `SUNContext` is `NULL` and any use of it otherwise should be avoided. If there is no error handler attached, errors will be ignored (by design).

File: src/sundials/sundials_logger_impl.h [hunk 2]
```diff
@@ -59,4 +60,37 @@ struct SUNLogger_
   SUNErrCode (*destroy)(SUNLogger* logger);
 };
 
+static void sunCreateLogMessage(SUNLogLevel lvl, int rank, const char* scope,
```
Comment (gardner48) [inline src/sundials/sundials_logger_impl.h:63]: Does this need to be define in the header?

File: src/sundials/sundials_mpi_errors.c [hunk 3]
```diff
@@ -18,7 +18,7 @@
 #include <sundials/sundials_mpi_types.h>
 #include <unistd.h>
 
-static inline char* combineFileAndLine(int line, const char* file)
+static inline char* sunCombineFileAndLine(int line, const char* file)
```
Comment (gardner48) [inline src/sundials/sundials_mpi_errors.c:21]: Why not use the version from `sundials_utils.h`?

**Review by gardner48 — CHANGES_REQUESTED** [review]

File: scripts/format.sh [hunk 4]
```diff
@@ -13,11 +13,13 @@
...
 #
 # Usage:
-#    ./format.sh <path to directory to format>
+#    ./format.sh <paths to directories or files to format>
 # 
 # We require clang-format 17.0.4. Other versions may produce different styles!
 # ---------------------------------------------------------------------------------
 
-find $1 -iname '*.h' -o -iname '*.hpp' -o \
+paths="$@"
```
Comment (gardner48) [inline scripts/format.sh:21]: In `tidy_and_format.sh` if you reverse the input order so it's build path then directories format that would make it easier to pass multiple path to that script as well if that's helpful.

File: scripts/format.sh [hunk 5]
```diff
@@ -13,11 +13,13 @@
...
+
+find $paths -iname '*.h' -o -iname '*.hpp' -o \
```
Comment (gardner48) [inline scripts/format.sh:23]: shellcheck suggests something like (I added the input check because I forgot when using the script)
```suggestion
if [ $# -lt 1 ]; then
    echo "ERROR: At least one path to format required"
    exit 1
fi

paths=( "$@" )

find "${paths[@]}" -iname '*.h' -o -iname '*.hpp' -o \
```

File: src/sundials/fmod/CMakeLists.txt [hunk 6]
```diff
@@ -19,8 +19,6 @@ set(sundials_SOURCES
   fsundials_adaptcontroller_mod.f90
   fsundials_context_mod.c
   fsundials_context_mod.f90
-  fsundials_errors_mod.c
-  fsundials_errors_mod.f90
```
[the remark this answers is not in the document]
Comment (balos1) [reply src/sundials/fmod/CMakeLists.txt:23]: Yes, thats right. I wanted to avoid forcing users to have to `use` yet another module. I actually think we should combine all of the core modules into one to mimic how you would link to the libraries or include headers.  I have a branch where it is 95% done.

File: scripts/format.sh (hunk 4 above)
Comment (balos1) [reply scripts/format.sh:21]: I actually am going to remove `tidy_and_format.sh` because the clang-tidy part doesn't work too well (I had to do a lot of extra stuff to make it work). The easier way to apply clang-tidy is by enabling it in CMake (in the formatting PR I added this option too).

File: .git-blame-ignore-revs [hunk 7]
```diff
@@ -0,0 +1,2 @@
+78770ef9b40fb163bf6d0fd582b6a449f0298585
+e75d0c03345b230def5713cda2e1b2748171634b
```
[the remark this answers is not in the document]
Comment (balos1) [reply .git-blame-ignore-revs:2]: The second hash is for the `realtype --> sunrealtype` change  (e.g.  and the change for the related macros e.g. `RCONST`).

File: include/sundials/priv/sundials_errors_impl.h [hunk 8]
```diff
@@ -27,6 +27,10 @@
 #include "sundials/sundials_logger.h"
 #include "sundials/sundials_types.h"
 
+#ifdef __cplusplus /* wrapper to enable C++ usage */
+extern "C" {
+#endif
```
[the remark this answers is not in the document]
Comment (balos1) [reply include/sundials/priv/sundials_errors_impl.h:32]: Since enums do not have linkage, it does not matter.

[the remark this answers is not in the document]
Comment (balos1) [reply include/sundials/priv/sundials_errors_impl.h:32]: I moved the end of the extern to the bottom though, just for clarity.

File: src/sundials/sundials_logger_impl.h (hunk 2 above)
Comment (balos1) [reply src/sundials/sundials_logger_impl.h:63]: Need to be - no, but it doesn't hurt for this utility I don't think.

File: src/sundials/sundials_mpi_errors.c (hunk 3 above)
Comment (balos1) [reply src/sundials/sundials_mpi_errors.c:21]: It should use it

File: src/sundials/sundials_logger_impl.h (hunk 2 above)
Comment (gardner48) [reply src/sundials/sundials_logger_impl.h:63]: It doesn't hurt but we generally don't define things in a header unless they need to be there (i.e., to inline) here and this was in a `.c` file before

Comment (balos1) [reply src/sundials/sundials_logger_impl.h:63]: Oops, I read this as `inline` already. It would need `inline` to stay in the header. Ill move it.

File: include/sundials/priv/sundials_errors_impl.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (balos1) [reply include/sundials/priv/sundials_errors_impl.h:32]: > Looking through the headers I noticed `sundials_context.h` still includes `sundials/sundials_context.hpp`, I thought that was removed in an earlier PR.

Looks like we [deferred it](https://github.com/LLNL/sundials/pull/360#issuecomment-1806533562) to the "error handling PR". Making this change does require updating a few dozen examples though, this is why the build is failing now.

File: examples/cvode/hip/cvAdvDiff_kry_hip.cpp [hunk 9]
```diff
@@ -41,7 +41,8 @@
 #include <nvector/nvector_hip.h>
 #include <stdio.h>
 #include <stdlib.h>
-#include <sundials/sundials_math.h>    /* definition of ABS and EXP   */
+#include <sundials/sundials_core.hpp>
+#include <sundials/sundials_core.hpp>  /* definition of ABS and EXP   */
```
Comment (gardner48) [inline examples/cvode/hip/cvAdvDiff_kry_hip.cpp:45]: Included twice
```suggestion
```

File: doc/superbuild/source/developers/style_guide/SourceCode.rst [hunk 10]
```diff
@@ -395,46 +338,81 @@ Coding Conventions and Rules
...
+         .doSomething()       //
+         .doAnotherThing()    //
+         .doSomethingElse();  //
+
+There are other scenarios (e.g., a function call with a lot of parameters) where doing this type of line break is useful for readability too.
+
+..
+   .. code-block::
+..    cd scripts
+..    ./tidy_and_format.sh <path to directory to tidy and format> <path to builddir>
```
Comment (gardner48) [inline doc/superbuild/source/developers/style_guide/SourceCode.rst:384]: `tidy_and_format.sh` was removed
```suggestion
..    ./format.sh <path to directory to tidy and format> <path to builddir>
```

File: doc/superbuild/source/developers/style_guide/SourceCode.rst [hunk 11]
```diff
@@ -395,46 +338,81 @@ Coding Conventions and Rules
...
+.. code-block:: cpp
+
+   MyObject::callAFunctionOfSorts()
+         .doSomething()       //
+         .doAnotherThing()    //
+         .doSomethingElse();  //
+
+There are other scenarios (e.g., a function call with a lot of parameters) where doing this type of line break is useful for readability too.
+
+..
```
Comment (gardner48) [inline doc/superbuild/source/developers/style_guide/SourceCode.rst:381]: Did you intend for the text below to be commented out?

File: scripts/tidy_and_format.sh
Comment (gardner48) [inline scripts/tidy_and_format.sh:1]: Why was this removed? Was there an issue with using `clang-format`? Do the `GitHub` actions need to be updated?

**Review by gardner48 — CHANGES_REQUESTED** [review]
Minor stuff

File: scripts/tidy_and_format.sh
Comment (gardner48) [reply scripts/tidy_and_format.sh:1]: I an earlier comment you said that the clang-tidy part didn't work too well. Did you mean clang-format?

Comment (balos1) [reply scripts/tidy_and_format.sh:1]: `clang-format` - formats the code
`clang-tidy` - lints the code, I used it to add missing braces and nothing else

The GitHub action uses `format.sh` which just does `clang-format` only, so no the action does not need to be updated.

I can explain in more detail about the difficulties of using `clang-tidy` offline.

File: doc/superbuild/source/developers/style_guide/SourceCode.rst (hunk 11 above)
Comment (balos1) [reply doc/superbuild/source/developers/style_guide/SourceCode.rst:381]: Yes, I did not want to lose what I had wrote but I also did not want it visible in the developer docs at this moment.

**Review by gardner48 — APPROVED** [review]

**Comment by balos1** [comment]
OK, now that this is approved, I am going to close it since the stuff is already in feature/error-handling-staging.
